File: smash-ranks-android/app/src/main/java/com/garpr/android/preferences/persistent/BasePersistentPreference.java
package com.garpr.android.preferences.persistent;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.garpr.android.preferences.BasePreference;
import com.garpr.android.preferences.KeyValueStore;

public abstract class BasePersistentPreference<T> extends BasePreference<T> {

    private final KeyValueStore mKeyValueStore;


    public BasePersistentPreference(@NonNull final String key,
            @Nullable final T defaultValue, @NonNull final KeyValueStore keyValueStore) {
        super(key, defaultValue);
        mKeyValueStore = keyValueStore;
    }

    @Override
    public void delete(final boolean notifyListeners) {
        mKeyValueStore.remove(getKey());

        if (notifyListeners) {
            notifyListeners();
        }
    }

    @Override
    public boolean exists() {
        return hasValueInStore() || getDefaultValue() != null;
    }

    protected KeyValueStore getKeyValueStore() {
        return mKeyValueStore;
    }

    protected boolean hasValueInStore() {
        return mKeyValueStore.contains(getKey());
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/preferences/persistent/BasePersistentPreference.kt
package com.garpr.android.preferences.persistent

import com.garpr.android.preferences.BasePreference
import com.garpr.android.preferences.KeyValueStore

abstract class BasePersistentPreference<T : Any>(
        key: String,
        defaultValue: T?,
        protected val keyValueStore: KeyValueStore
) : BasePreference<T>(
        key,
        defaultValue
) {

    override val exists: Boolean
        get() = hasValueInStore || defaultValue != null

    protected val hasValueInStore: Boolean
        get() = key in keyValueStore

    final override fun performDelete() {
        keyValueStore.remove(key)
    }

}
